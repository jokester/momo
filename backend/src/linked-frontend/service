../../submodules/momocraft-frontend/src/service