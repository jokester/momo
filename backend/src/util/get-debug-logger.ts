import path from 'path';
import debug from 'debug';
import { getDebugNamespace } from '@jokester/ts-commonutil/cjs/logging/get-debug-namespace';

const createNamespace = getDebugNamespace(path.join(__dirname, '..'));
export function getDebugLogger(srcFile: string): debug.IDebugger {
  const namespace = 'hanko:' + createNamespace(srcFile);
  return debug(namespace);
}
