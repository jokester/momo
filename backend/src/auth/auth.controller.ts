import { BadRequestException, Body, Controller, Get, Post, Req, Request, UseGuards } from '@nestjs/common';
import { AuthService } from './auth.service';
import { UserEntity } from '../entity/user.entity';
import { LocalStrategyGuard } from './local.strategy';
import { getDebugLogger } from '../util/get-debug-logger';
import { GoogleOAuthService } from './google-oauth.service';
import { UserService } from '../user/user.service';

const logger = getDebugLogger(__filename);

@Controller('auth')
export class AuthController {
  constructor(
    private readonly authService: AuthService,
    private readonly googleOAuthService: GoogleOAuthService,
    private readonly userService: UserService,
  ) {}

  @UseGuards(LocalStrategyGuard)
  @Post('email/signin')
  async emailSignIn(@Request() req: Request & { user?: UserEntity }): Promise<UserEntity | null> {
    logger('got email user: %o', req.user);
    return req.user || null;
  }

  @Get('google/oauth' /* ?code=callback_code... */)
  async startGoogleOAuth(@Request() req: Request & { user?: UserEntity }) {
    logger('got google/oauth user: %o', req.user);
    return null;
  }

  @Get('ping')
  async pingDb() {
    await this.userService.now();
    return null;
  }

  @Post('oauth/google')
  async doGoogleOAuth(@Request() req: Request, @Body() payload2: { code: string; callbackUri: string }) {
    // const payload: { code: string; callbackUri: string } = await req.json();

    const payload = payload2;
    if (payload && payload.code) {
      const x = await this.googleOAuthService.auth(payload.code);

      logger('got auth', x);
      return x;
    }
    throw new BadRequestException();
  }
}
