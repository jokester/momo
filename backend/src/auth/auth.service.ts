import { Injectable } from '@nestjs/common';
import { getDebugLogger } from '../util/get-debug-logger';

@Injectable()
export class AuthService {
  constructor() {
    getDebugLogger(__filename)('authService created');
  }

  touch() {
    getDebugLogger(__filename)('authService touched');
  }
}
