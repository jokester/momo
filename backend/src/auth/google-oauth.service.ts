import { google } from 'googleapis';
import { Injectable } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { getDebugLogger } from '../util/get-debug-logger';
import { oauth2 } from 'googleapis/build/src/apis/oauth2';

const logger = getDebugLogger(__filename);

@Injectable()
export class GoogleOAuthService {
  constructor(private readonly confService: ConfigService) {}

  async auth(code: string) {
    const client = this.createAuthClient();
    logger('client=%O', client);
    logger('code=', code);

    const token = await client.getToken(code);
    client.setCredentials(token.tokens);
    logger('token=%O', token);

    const x = await oauth2({ version: 'v2', auth: client });
    const got = await x.userinfo.get();
    return got.data;
  }

  private createAuthClient() {
    return new google.auth.OAuth2({
      clientId: this.confService.get('OAUTH_GOOGLE_CLIENT_ID'),
      clientSecret: this.confService.get('OAUTH_GOOGLE_CLIENT_SECRET'),
      redirectUri: this.confService.get('OAUTH_GOOGLE_REDIRECT_URI'),
    });
  }
}
