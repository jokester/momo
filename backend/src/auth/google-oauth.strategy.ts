import { AuthGuard, PassportStrategy } from '@nestjs/passport';
import { Profile, Strategy } from 'passport-google-oauth20';
import { UserEntity } from '../entity/user.entity';
import { Injectable, InternalServerErrorException } from '@nestjs/common';
import { getDebugLogger } from '../util/get-debug-logger';

@Injectable()
export class GoogleOauthStrategy extends PassportStrategy(Strategy, 'google') {
  constructor() {
    super({
      clientID: 'FILL ME',
      clientSecret: 'FILL_ME',
      scope: ['profile', 'email'],
      callbackURL: 'FILL_ME', // have to match the one retrieved code?
    });
  }

  async validate(accessToken: string, refreshToken: string, profile: Profile): Promise<UserEntity> {
    getDebugLogger(__filename)('%s / %s / %O', accessToken, refreshToken, profile);

    throw new InternalServerErrorException('something is wrong');
  }
}

export const GoogleOAuthGuard = AuthGuard('google');
