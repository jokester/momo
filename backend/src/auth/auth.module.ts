import { Module } from '@nestjs/common';
import { AuthController } from './auth.controller';
import { DiscordOAuth, GoogleOAuth } from './oauth-client.provider';
import { DiscordOAuthService } from './discord-oauth.service';
import { GoogleOAuthService } from './google-oauth.service';
import { UserModule } from '../user/user.module';

@Module({
  imports: [UserModule],
  controllers: [AuthController],
  providers: [DiscordOAuth.Provider, DiscordOAuthService, GoogleOAuth.Provider, GoogleOAuthService],
})
export class AuthModule {}
