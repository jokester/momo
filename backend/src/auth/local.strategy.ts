import { AuthGuard, PassportStrategy } from '@nestjs/passport';
import { Strategy } from 'passport-local';
import { Injectable, UnauthorizedException } from '@nestjs/common';
import { AuthService } from './auth.service';
import { getDebugLogger } from '../util/get-debug-logger';
import { UserEntity } from '../entity/user.entity';

@Injectable()
export class LocalStrategy extends PassportStrategy(Strategy) {
  constructor(private readonly authService: AuthService) {
    super();
    getDebugLogger(__filename)('gotAuthService: %o', authService);
  }

  async validate(username: string, password: string): Promise<UserEntity> {
    const user = await this.authService.authUserWithPassword(username, password);

    if (user.isSome()) {
      return user.value;
    }
    throw new UnauthorizedException();
  }
}

export const LocalStrategyGuard = AuthGuard('local');
