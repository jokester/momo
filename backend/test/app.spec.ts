import request from 'supertest';
import { Test } from '@nestjs/testing';
import { AppModule } from '../src/app.module';
import { INestApplication } from '@nestjs/common';
import { GoogleOAuthResponse, GoogleOAuthService } from '../src/auth/google-oauth.service';
import { MockData, TestDeps } from './test-deps';
import { right } from 'fp-ts/lib/Either';
import { AuthModule } from '../src/auth/auth.module';
import { TypeORMConnection } from '../src/db/typeorm-connection.provider';
import { JwtService } from '@nestjs/jwt';
import { getDebugLogger } from '../src/util/get-debug-logger';

const logger = getDebugLogger(__filename);

describe('AppController (e2e)', () => {
  let app: INestApplication;
  let jwtService: JwtService;

  beforeAll(async () => {
    const moduleFixture = await Test.createTestingModule({
      imports: [AppModule],
    })
      .overrideProvider(TypeORMConnection)
      .useValue(TestDeps.testConnection)
      .compile();

    app = moduleFixture.createNestApplication();
    jwtService = await moduleFixture.resolve(JwtService);

    await app.init();
    await TestDeps.clearTestDatabase();
  });

  afterAll(async () => {
    await app.close();
  });

  describe(JwtService, () => {
    const origPayload = { something: 'really something' } as const;
    it('signs payload to jwt token', async () => {
      const token = await jwtService.signAsync(origPayload);
      const verified = await jwtService.verify<typeof origPayload>(token);

      expect(verified.something).toEqual(origPayload.something);
    });

    it('throws on different key', async () => {
      const anotherJwtService = new JwtService({ secret: 'aaaasecret' });
      const token = await anotherJwtService.signAsync(origPayload);
      await expect(jwtService.verifyAsync(token)).rejects.toThrowError('invalid signature');
    });

    it('throws on malformed jwt token', async () => {
      const token = 'ajfal;sdf';
      expect(() => jwtService.verify(token)).toThrowError('jwt malformed');
    });

    it('throws on expired token', async () => {
      const token = await jwtService.signAsync(origPayload);
      await expect(
        jwtService.verifyAsync(token, { clockTimestamp: Date.now() / 1e3 + 7 * 24 * 3600 - 10 }),
      ).resolves.toBeTruthy();

      await expect(
        jwtService.verifyAsync(token, { clockTimestamp: Date.now() / 1e3 + 7 * 24 * 3600 }),
      ).rejects.toThrowError(/jwt expired/);
    });
  });

  describe(AuthModule, () => {
    it('POST /auth/oauth/google', async () => {
      jest
        .spyOn(GoogleOAuthService.prototype, 'auth')
        .mockResolvedValue(right<string, GoogleOAuthResponse>(MockData.googleOAuthResponseValid));

      const res = await request(app.getHttpServer())
        .post('/auth/oauth/google')
        .send({ code: '123', redirectUrl: '456' })
        .expect(201);

      const jwtToken = JSON.parse(res.text).jwtToken;
      expect(jwtToken).toBeTruthy();

      const decoded = await jwtService.decode(jwtToken);
      logger('decoded', decoded);

      await jwtService.verifyAsync(jwtToken);
    });
  });
});
