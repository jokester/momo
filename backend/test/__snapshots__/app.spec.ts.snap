// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppController (e2e) AuthModule POST /auth/oauth/google returns jwtToken on succeed: jwt/validate 1`] = `
Object {
  "avatarUrl": "https://example.com/a.png",
}
`;
